### PR TITLE
Enable support for count() method for unicode string

### PR DESCRIPTION
**Pull request by Vyacheslav-Smirnov**
Changes are done on the head of https://github.com/numba/numba/pull/4204 PR
 - Update documentation
 - Rewrite the algorithm
 - Refactor tests for count

### CODE REVIEW

**Comment by Vyacheslav-Smirnov** [comment]
Hi, please set the 'In progress' label for this PR
@shssf , could you please review the changes?

File: numba/tests/test_unicode.py [hunk 1]
```diff
@@ -7,6 +7,7 @@
 
 import sys
 from itertools import permutations
+from itertools import product
```
Comment (shssf) [inline numba/tests/test_unicode.py]: move to previous line

File: numba/tests/test_unicode.py [hunk 2]
```diff
@@ -338,6 +362,68 @@ def test_find(self, flags=no_pyobj_flags):
...
                                  "'%s'.find('%s')?" % (a, substr))
 
+    def test_count(self):
+        pyfunc = count_usecase
+        cfunc = njit(pyfunc)
+
+        for s in UNICODE_REPEATING_EXAMPLES:
+            extras = ['', ' ', 'xx', s[::-1], s[:-2], s[3:], s, s + s,
+                      'aa', 'ab', 'bb', 'aaa', 'aba']
+            for sub in extras:
```
Comment (shssf) [inline numba/tests/test_unicode.py]: Create `UNICODE_REPEATING_EXAMPLES ` as `STRIP_CASES` and use second field instead `extras` here. Non need to run algorithm on all combinations.

File: numba/tests/test_unicode.py [hunk 3]
```diff
@@ -338,6 +362,68 @@ def test_find(self, flags=no_pyobj_flags):
...
+        with self.assertRaises(TypingError) as raises:
+            cfunc('ascii', 'c', 1, 0.5)
+        self.assertIn('slice indices must be integers or None', str(raises.exception))
+        with self.assertRaises(TypingError) as raises:
+            cfunc('abcde', 's', 1.2, 7)
+        self.assertIn('slice indices must be integers or None', str(raises.exception))
+        with self.assertRaises(TypingError) as raises:
+            cfunc('abcde', 's', 'ab', 7)
+        self.assertIn('slice indices must be integers or None', str(raises.exception))
+        with self.assertRaises(TypingError) as raises:
```
Comment (shssf) [inline numba/tests/test_unicode.py]: Could you please use empty line between logical algorithm blocks?

File: numba/unicode.py [hunk 4]
```diff
@@ -440,6 +440,14 @@ def _codepoint_is_ascii(ch):
     return ch < 128
 
 
+@register_jitable
+def normalize_slice_arg(arg, slice_len, default):
+    if arg is None:
```
Comment (shssf) [inline numba/unicode.py]: Documentation?
Is this function do something else than `normalize_str_idx`?

File: numba/unicode.py [hunk 5]
```diff
@@ -540,6 +548,40 @@ def find_impl(a, b):
         return find_impl
 
 
+@overload_method(types.UnicodeType, 'count')
+def unicode_count(src, sub, start=None, end=None):
+
+    count_args_types_check(start)
+    count_args_types_check(end)
+
+    if isinstance(sub, types.UnicodeType):
```
Comment (shssf) [inline numba/unicode.py]: Can `sub` be other type like `types.UnicodeCharSeq` (Optional, None)?

Comment (Vyacheslav-Smirnov) [reply numba/unicode.py]: As I understand, it should be UnicodeType, due to UnicodeCharSeq does not match/cast cprreclty to unicode string (as mentioned in there: https://github.com/numba/numba/issues/4018)

File: numba/tests/test_unicode.py [hunk 6]
```diff
@@ -338,6 +396,61 @@ def test_find(self, flags=no_pyobj_flags):
...
+        self.assertIn('slice indices must be integers or None', str(raises.exception))
+
+        with self.assertRaises(TypingError) as raises:
+            cfunc('abcde', 's', 'ab', 7)
+        self.assertIn('slice indices must be integers or None', str(raises.exception))
+
+        with self.assertRaises(TypingError) as raises:
+            cfunc('abcde', 's', 1, 'ab')
+        self.assertIn('slice indices must be integers or None', str(raises.exception))
+
```
[the remark this answers is not in the document]
Comment (stuartarchibald) [reply numba/tests/test_unicode.py]: Similarly optional types are not exercised.

**Comment by Vyacheslav-Smirnov** [comment]
Hi, thank you for the comments
Here is changes for tests. I am also going to refactor the algorithm in few next hours - need to remove slice using to avoid string copying

File: numba/unicode.py [hunk 7]
```diff
@@ -944,6 +987,26 @@ def normalize_str_idx(idx, length, is_start=True):
...
+def _normalize_slice_idx_count(arg, slice_len, default):
+    """
+    Used for unicode_count
+
+    If arg < -slice_len, returns 0 (prevents circle)
+
+    If arg is within slice, e.g -slice_len <= arg < slice_len
+    returns its real index via arg % slice_len
+
+    If arg > slice_len, returns arg (in this case count must return 0 if it is srart index)
```
Comment (stuartarchibald) [inline numba/unicode.py]:
```suggestion
    If arg > slice_len, returns arg (in this case count must return 0 if it is start index)
```

**Comment by stuartarchibald** [comment]
> Hi, thank you for the comments
> Here is changes for tests. I am also going to refactor the algorithm in few next hours - need to remove slice using to avoid string copying

Are the slices not returning views? There's no use of slice with a step in the alg and IIRC the bounds are normalized, so I'd anticipate this code being used https://github.com/numba/numba/pull/4606/files#diff-a21787256ad779b4c4c5c99915c71025R1106-R1112 ? If for some reason it is not being used, I'd prefer to resolve this through improving the use of slices as views than adding complexity to an algorithm, largely because the improvement to slices would likely be applicable to many cases.

File: numba/unicode.py (hunk 7 above)
Comment (stuartarchibald) [inline numba/unicode.py]:
```suggestion
    If arg > slice_len, returns arg (in this case count must return 0 if it is start index)
```

Comment (stuartarchibald) [reply numba/unicode.py]: also, all lines need to be <80 chars.

**Comment by stuartarchibald** [comment]
Thanks for making the updates in https://github.com/numba/numba/pull/4606/commits/23d09fb77e900449f100e69229f8b878c4c1e419, much appreciated.

**Comment by Vyacheslav-Smirnov** [comment]
Ok, so @stuartarchibald  am I understand right that the least is to make code-style changes (all lines need to be <80 chars) and the other part (the algorithm) is good for you?

**Comment by stuartarchibald** [comment]
> Ok, so @stuartarchibald am I understand right that the least is to make code-style changes (all lines need to be <80 chars) and the other part (the algorithm) is good for you?

All code should be formatted to <80 chars. The algorithm in this PR looks fine to me, unless you have specific and demonstrable concerns about performance, in which case we should try and resolve those. Tests look good, with the exception of the unresolved comments I made above that need fixes. Hopefully it's not to much work to fix these few things and this can make the release. Thanks again.

**Comment by Vyacheslav-Smirnov** [comment]
@stuartarchibald , Thank you for comments, I have done with the changes.

File: numba/tests/test_unicode.py [hunk 8]
```diff
@@ -338,6 +386,110 @@ def test_find(self, flags=no_pyobj_flags):
...
+                      str(raises.exception))
+
+        with self.assertRaises(TypingError) as raises:
+            cfunc('ascii', 'c', 1.2, 7)
+        self.assertIn('The slice indices must be an Integer or None',
+                      str(raises.exception))
+
+        with self.assertRaises(TypingError) as raises:
+            cfunc('ascii', 12, 1.2, 7)
+        self.assertIn('The slice indices must be an Integer or None',
```
Comment (stuartarchibald) [inline numba/tests/test_unicode.py]: This isn't the first problem though is it? It is a problem, but the real issue is trying to `'ascii'.count(12)`?

Comment (stuartarchibald) [reply numba/tests/test_unicode.py]: oh I see, this emulates python, slice raises ahead of type check on the counted string.

Comment (stuartarchibald) [reply numba/tests/test_unicode.py]: It'd be nice if this:
```
cfunc('ascii', 12, 1, 7)
```
raised with an error message that was better than:
```
*** numba.errors.TypingError: Failed in nopython mode pipeline (step: nopython frontend)
Invalid use of BoundFunction((<class 'numba.types.misc.UnicodeType'>, 'count') for unicode_type) with parameters (int64, int64, int64)
```

**Comment by stuartarchibald** [comment]
@Vyacheslav-Smirnov thanks for the fix ups, there's just this one usability case https://github.com/numba/numba/pull/4606#discussion_r329017323 that I think ought to be addressed and then this is good to go.

I presume you were happy with the `slice` performance?

Thanks.

**Comment by Vyacheslav-Smirnov** [comment]
Add the special Typing error for  substring.
Regarding the slice performance, need to test with different realizations to check the performance, but think that this can be done in scope of other PR/activity.

**Comment by stuartarchibald** [comment]
> Add the special Typing error for substring.

Great, thanks very much.

> Regarding the slice performance, need to test with different realizations to check the performance, but think that this can be done in scope of other PR/activity.

Agree. Thanks.

**Comment by stuartarchibald** [comment]
@Vyacheslav-Smirnov Congratulations on your first contribution to Numba :)

**Comment by Vyacheslav-Smirnov** [comment]
Thank you! I am pleased to work with you)

**Pull request merged**